modules/navier_stokes/src/postprocessors/VolumetricFlowRate.C

```diff
@@ -16,23 +16,91 @@ InputParameters
 VolumetricFlowRate::validParams()
 {
   InputParameters params = SideIntegralPostprocessor::validParams();
-  params.addClassDescription("Computes the volumetric flow rate through a boundary.");
-  params.addRequiredCoupledVar("vel_x", "The x velocity");
-  params.addCoupledVar("vel_y", 0, "The y velocity");
-  params.addCoupledVar("vel_z", 0, "The z velocity");
+  params.addClassDescription(
+      "Computes the volumetric flow rate of an advected quantity through an external boundary.");
+  params.addParam<bool>("fv", false, "Whether finite volume variables are used");
+  params.addRequiredCoupledVar("vel_x", "The x-axis velocity");
+  params.addCoupledVar("vel_y", 0, "The y-axis velocity");
+  params.addCoupledVar("vel_z", 0, "The z-axis velocity");
+  params.addCoupledVar(
+      "advected_variable", 0, "The advected variable quantity of which to study the flow");
+  params.addParam<MaterialPropertyName>(
+      "advected_mat_prop", 0, "The advected material property of which to study the flow");
   return params;
 }
 
 VolumetricFlowRate::VolumetricFlowRate(const InputParameters & parameters)
   : SideIntegralPostprocessor(parameters),
+    _fv(getParam<bool>("fv")),
     _vel_x(coupledValue("vel_x")),
     _vel_y(coupledValue("vel_y")),
-    _vel_z(coupledValue("vel_z"))
+    _vel_z(coupledValue("vel_z")),
+    _fv_vel_x(dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar("vel_x", 0))),
+    _fv_vel_y(dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar("vel_y", 0))),
+    _fv_vel_z(dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar("vel_z", 0))),
+    _advected_variable(coupledValue("advected_variable")),
+    _fv_advected_variable(
+        dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar("advected_variable", 0))),
+    _advected_material_property(getADMaterialProperty<Real>("advected_mat_prop"))
 {
+  /// Check that at most one advected quantity has been provided
+  mooseAssert(!parameters.isParamSetByUser("advected_variable") ||
+                  !parameters.isParamSetByUser("advected_mat_prop"),
+              "VolumetricFlowRatePostprocessor should be provided either an advected variable or "
+              "an advected material property");
 }
 
 Real
 VolumetricFlowRate::computeQpIntegral()
 {
-  return RealVectorValue(_vel_x[_qp], _vel_y[_qp], _vel_z[_qp]) * _normals[_qp];
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+  if (_fv)
+  {
+    /// We should be at the edge of the domain
+    const FaceInfo * const fi = _mesh.faceInfo(_current_elem, _current_side);
+    mooseAssert(fi, "We should have a face info");
+    mooseAssert(fi->isBoundary(), "VolumetricFlowRate should only be used at boundaries");
+
+    /// Get face value for velocity
+    const auto & vx_face = !getFieldVar("vel_x", 0)
+                               ? _vel_x[_qp]
+                               : MetaPhysicL::raw_value(_fv_vel_x->getBoundaryFaceValue(*fi));
```

Comment:
in case you're wondering this whole shenanigan is to capture the case where a user would set "vel_x = 2" instead of a variable